ert3/storage/_storage.py

```diff
@@ -96,45 +111,147 @@ def get_experiment_names(*, workspace: Path) -> Set[str]:
+        )
+
+        if meta_response.status_code != 200:
+            raise ert3.exceptions.StorageError(meta_response.text)
+
+
+def _response2record(
+    response_content: bytes, record_type: ert3.data.RecordType, realization_id: int
+) -> ert3.data.Record:
+    dataframe = pd.read_csv(
+        io.BytesIO(response_content), index_col=0, float_precision="round_trip"
     )
-    if response.status_code == 409:
-        raise KeyError("Record already exists")
+
+    raw_index = tuple(dataframe.columns)
+    if record_type == ert3.data.RecordType.LIST_FLOAT:
+        array_data = tuple(
+            float(dataframe.loc[realization_id][raw_idx]) for raw_idx in raw_index
+        )
+        return ert3.data.Record(data=array_data)
+    elif record_type == ert3.data.RecordType.MAPPING_INT_FLOAT:
+        int_index = tuple(int(e) for e in dataframe.columns)
+        idata = {
+            idx: float(dataframe.loc[realization_id][raw_idx])
+            for raw_idx, idx in zip(raw_index, int_index)
+        }
+        return ert3.data.Record(data=idata)
+    elif record_type == ert3.data.RecordType.MAPPING_STR_FLOAT:
+        str_index = tuple(str(e) for e in dataframe.columns)
+        sdata = {
+            idx: float(dataframe.loc[realization_id][raw_idx])
+            for raw_idx, idx in zip(raw_index, str_index)
+        }
+        return ert3.data.Record(data=sdata)
+    else:
+        raise ValueError(
+            f"Unexpected record type when loading numerical record: {record_type}"
+        )
+
+
+def _get_numerical_metadata(ensemble_id: str, record_name: str) -> _NumericalMetaData:
+    response = requests.get(
+        url=f"{_STORAGE_URL}/ensembles/{ensemble_id}/records/{record_name}/metadata",
+        params={"realization_index": 0},  # This assumes there is a realization 0
+    )
+
+    if response.status_code == 404:
+        raise ert3.exceptions.ElementMissingError(
+            f"No metadata for {record_name} in ensemble: {ensemble_id}"
+        )
 
     if response.status_code != 200:
         raise ert3.exceptions.StorageError(response.text)
 
+    return _NumericalMetaData(**json.loads(response.content))
+
 
-def _get_data(workspace: Path, experiment_name: str, record_name: str) -> Any:
+def _get_numerical_data(
+    workspace: Path, experiment_name: str, record_name: str
+) -> ert3.data.EnsembleRecord:
     experiment = _get_experiment_by_name(experiment_name)
     if experiment is None:
-        raise KeyError(
+        raise ert3.exceptions.NonExistantExperiment(
             f"Cannot get {record_name} data, no experiment named: {experiment_name}"
         )
 
     ensemble_id = experiment["ensembles"][0]  # currently just one ens per exp
-    response = requests.get(
-        url=f"{_STORAGE_URL}/ensembles/{ensemble_id}/records/{record_name}"
-    )
+    metadata = _get_numerical_metadata(ensemble_id, record_name)
+
+    records = []
+    for real_id in range(metadata.ensemble_size):
+        response = requests.get(
```

Comment:
This is going to be excruciatingly slow. I suggest `aiohttp` unless there's a batch endpoint in storage…

Reply:
Ignore if you didn't introduce this.

Reply:
I did introduce it, as earlier the entire ensemble record was posted as a single file into storage. However, the main goal of this PR was not to gain speed, but to start posting the data as numerical data.

I listed optimisation/speed under future work. However, there is a batch endpoint that we can try to use. After a quick examination of our GitHub Actions it seems like this PR brings the joint runtime of all the polynomial experiments from 4 to 8 minutes. Which is a significant slowdown 🤔

But, for real speed I'm suspecting that we need to expand and refine the notion of a record both in ert and in ert-storage. Because, currently we will not be able to get away from iterating over all of the data using Python. And that is going to be somewhat slooow no matter how we do the rest...

I can have a look at whether the batch endpoint helps alone, but I still think we will need to address efficiency down the road on a more holistic level. And that I hope we agree upon is beyond the scope of this PR :)

Reply:
After further consideration I suggest that we keep this as is... Because this code should ASAP be moved into a storage record transmitter and then it will be done as sketched here, but for the different realisations individually :)